tools/builder/systemd-builder.py

```diff
@@ -0,0 +1,170 @@
+        # on a hypervisor (such as Docker for Mac) will bind-mount external
+        # volumes as root inside the container.
+        #
+        # Also when we use a scratch volume for the build, the volume has
+        # root ownership.
+        #
+        # This is not ideal, but proceeding here is still a pragmatic solution
+        # to handle these issues.
+        #
+        # Do not reexec_unprivileged in this case.
+        return
+    if uid < 1000 or gid < 1000:
+        raise Error('Build tree UID/GID outside of user range: {}/{}'.format(uid, gid))
+    if not check_builder_user_exists(uid, gid):
+        create_builder_user(uid, gid)
+    reexec_unprivileged(uid, gid)
+
+# Argument parser: Implement subcommands for operations. For now "build" is the
+# common one, but we also want to support "setup" for configure the
+# systemd-builder container.
+#
+# This can be further extended with specialized build types, such as
+# "cov-build" for Coverity and potentially other specialized build
+# configurations.
+
+def create_parser():
+    parser = argparse.ArgumentParser(prog=PROGNAME)
+    subparsers = parser.add_subparsers(dest='cmd')
+    subparsers.add_parser('setup')
+    subparsers.add_parser('debugshell')
+    build_parser = subparsers.add_parser('build', help='build source tree')
+    build_parser.add_argument('--test', action='store_true', help='run ninja tests after building')
+    build_parser.add_argument('--meson_args', action='store', help='arguments to pass to meson')
+    build_parser.add_argument('--ninja_args', action='store', default='-v', help='arguments to pass to ninja')
+    build_parser.add_argument('--test_args', action='store', help='arguments to pass to ninja test')
+    return parser
+
+# Subcommands: These implement the actual commands such as "build" and "setup".
+# More commands can be added for specialized builders.
+
+def cmd_setup(opts):
+    must_run_as_root('setup')
+    os.mkdir(BUILDER_SRC_TREE)
+    os.mkdir(BUILDER_DST_TREE)
+
+def cmd_debugshell(opts):
+    sys.stdout.flush()
+    sys.stderr.flush()
+    os.execv('/bin/bash', ['-/bin/bash'])
+
+def check_src_tree():
+    if not os.path.ismount(BUILDER_SRC_TREE):
+        raise Error('Expected source tree [{}] to be a mount.'.format(BUILDER_SRC_TREE))
+    # Look for a well-known file in the source tree.
+    if not os.path.exists(os.path.join(BUILDER_SRC_TREE, 'src/core/main.c')):
+        raise Error('Source tree [{}] does not look like it contains systemd sources.'.format(BUILDER_SRC_TREE))
+
+def cmd_build(opts):
+    rerun_as_builder()
+    check_src_tree()
+    # Assemble the `meson` command-line.
+    meson_cmd = ['meson']
+    if os.path.exists(os.path.join(BUILDER_DST_TREE, 'build.ninja')):
+        # This tree was previously used, so run `meson configure` instead.
+        meson_cmd.append('configure')
+    meson_cmd.append(BUILDER_DST_TREE)
+    meson_cmd.extend(shlex.split(opts.meson_args))
```

Comment:
`shlex.split(None)` will try to read from stdin. `meson_args` and `test_args` should have a default of `''` (the empty string) when set above.